File: server/src/services/task.ts
import { priority, status, useDrizzle } from '@/config';
import type { PostgresJsDatabase } from 'drizzle-orm/postgres-js';
import { task } from '@/config';
import { eq } from 'drizzle-orm';
import { InternalError, ServiceSuccess } from '@/enums';
import type { IBaseService } from '@bunarcane/arcane';
import { InternalServerError, NotFoundError } from '@/errors';

interface ITaskService extends IBaseService<TTask, number> {
    findTaskByNode(nodeId: number): Promise<TOmits<TTask, 'nodeId' | 'createdAt'>>;
}

export class TaskService implements ITaskService {
    db: PostgresJsDatabase<Record<string, never>>;

    constructor() {
        this.db = useDrizzle();
    }

    list(): Promise<TTask[]> {
        throw new Error('Not implemented');
    }

    async findTaskByNode(nodeId: number): Promise<TOmits<TTask, 'nodeId' | 'createdAt'>> {
        return await this.db
            .select({
                id: task.id,
                title: task.title,
                description: task.description,
                startDate: task.startDate,
                dueDate: task.dueDate,
                color: task.color,
                minLength: task.minLength,
                maxLength: task.maxLength,
                priority: {
                    display: priority.display,
                    description: priority.description,
                    color: priority.color,
                },
                status: {
                    display: status.display,
                    description: status.description,
                    color: status.color,
                },
            })
            .from(task)
            .innerJoin(priority, eq(task.priorityId, priority.id))
            .innerJoin(status, eq(task.statusId, status.id))
            .where(eq(task.nodeId, nodeId));
    }

    async create(value: Omit<TTask, 'id'>): Promise<string> {
        const result = await this.db.insert(task).values(value).returning();
        if (result.empty()) throw new InternalServerError(InternalError('task').CREATE);
        return ServiceSuccess.CREATE;
    }

    async update(id: number, value: Omit<TTask, 'id'>): Promise<string> {
        const target = await this.db.select().from(task).where(eq(task.id, id));
        if (target.empty()) throw new NotFoundError(InternalError('task').NONEXISTENCE('id', id));

        const result = await this.db.update(task).set(value).where(eq(task.id, id)).returning();
        if (result.empty()) throw new InternalServerError(InternalError('task').UPDATE);

        return ServiceSuccess.UPDATE;
    }

    async remove(id: number): Promise<string> {
        const result = await this.db.delete(task).where(eq(task.id, id)).returning();
        if (result.empty()) throw new InternalServerError(InternalError('task').DELETE);
        return ServiceSuccess.DELETE;
    }
}

File: server/internal/services/task.go
package services

import (
	"fmt"
	success "sukejuuru/internal/enums"
	"sukejuuru/internal/models"
	"sukejuuru/internal/utils"

	"github.com/jmoiron/sqlx"
)

type taskService struct {
	db *sqlx.DB
}

func UseTask() *taskService {
	return &taskService{db: utils.UsePostgres()}
}

func (service taskService) FindAllByNode(nodeId int) ([]models.TaskRelations, error) {
	var tasks []models.TaskRelations
	query := `
		SELECT t.id, t.title, t.description, t.min_length, t.max_length, t.start_date, t.due_date, t.color,
			p.display, p.description, p.color,
			s.display, s.description, s.color
		FROM task t
		JOIN priority p ON t.priority_id = p.id
		JOIN status s on t.status_id = s.id
		WHERE t.node_id = $1
	`
	if err := service.db.Select(&tasks, query, nodeId); err != nil {
		return nil, err
	}
	return tasks, nil
}

func (service taskService) Create(values models.Task) (string, error) {
	query := `
		INSERT INTO task (title, description, min_length, max_length, start_date, due_date, created_at, color, node_id, priority_id, status_id)
		VALUES (:title, :description, :min_length, :max_length, :start_date, :due_date, :created_at, :color, :node_id, :priority_id, :status_id)
	`
	if _, err := service.db.NamedExec(query, values); err != nil {
		return "", err
	}
	return success.CREATED, nil
}

func (service taskService) Patch(id int, target string, value any) (string, error) {
	query := fmt.Sprintf(`
		UPDATE TABLE task 
		SET %s = $1
		WHERE id = $2
	`, target)
	if _, err := service.db.Exec(query, value, id); err != nil {
		return "", err
	}
	return success.UPDATED, nil
}

func (service taskService) Delete(id int) (string, error) {
	query := "DELETE TABLE task WHERE id = $1"
	if _, err := service.db.Exec(query, id); err != nil {
		return "", err
	}
	return success.DELETED, nil
}
